### PR TITLE
[ROCm] Use hipify_torch as a submodule for hipification on ROCm

### PR DESCRIPTION
**Pull request by lcskrishna**
This PR enables hipify_torch as a third party submodule to perform hipification on ROCm.

Carries on from where we left off in https://github.com/pytorch/pytorch/pull/80136, with changes to ensure that the `python setup.py develop` use-case works as well

cc: @jithunnair-amd 

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @albanD

### CODE REVIEW

**Comment by ezyang** [comment]
@jithunnair-amd are you taking point on this?

**Comment by jithunnair-amd** [comment]
> @jithunnair-amd are you taking point on this?

@ezyang Yes

**Comment by jithunnair-amd** [comment]
@lcskrishna Just moving this PR to draft until we resolve any CI issues etc., and we can then move it to Ready for Review for upstream

File: torch/utils/hipify/__init__.py [hunk 1]
```diff
@@ -1 +1,5 @@
-from .version import __version__
+from hipify_torch.version import __version__
```
Comment (jithunnair-amd) [inline torch/utils/hipify/__init__.py:1]: @lcskrishna Should we add an `import hipify_torch` here to satisfy the mypy import issue?

Comment (lcskrishna) [reply torch/utils/hipify/__init__.py:2]: Done.

**Comment by jithunnair-amd** [comment]
@lcskrishna The `lintrunner / linux-job` errors are easily fixable with spaces.
A lot of the CUDA failures seem unrelated to your PR, maybe needs a rebase?

**Comment by jithunnair-amd** [comment]
@alband/@seemethere looks like we'll need your help for this one: https://github.com/pytorch/pytorch/actions/runs/5113942775/jobs/9193695569

This PR is to move hipify to a submodule so it will involve a lot of LOC.

**Comment by jithunnair-amd** [comment]
@malfet @kit1980 As for failures such as https://github.com/pytorch/pytorch/actions/runs/5113942768/jobs/9194006586#step:14:32171,
it's caused by the code [here](https://github.com/pytorch/pytorch/blob/main/test/test_testing.py#L2173) that essentially tries to ensure every .py file in the Pytorch repo is importable. If we need to satisfy that requirement, we would need to install hipify_torch for CUDA builds as well, which seems unnecessary. So shall we add an exception in this test for hipify?

File: torch/utils/cpp_extension.py [hunk 2]
```diff
@@ -1484,7 +1486,7 @@ def _jit_compile(name,
         baton = FileBaton(os.path.join(build_directory, 'lock'))
         if baton.try_acquire():
             try:
-                with GeneratedFileCleaner(keep_intermediates=keep_intermediates) as clean_ctx:
+                with hipify_python.GeneratedFileCleaner(keep_intermediates=keep_intermediates) as clean_ctx:
```
Comment (jithunnair-amd) [inline torch/utils/cpp_extension.py:1489]: @lcskrishna We will need to figure out a way to:

1. either add a `catch` block for the error thrown by this line on CUDA builds, since `hipify_python` is set to `None` here: https://github.com/pytorch/pytorch/pull/101391/files#diff-258d5272916c8cf5ac89aad77f7c114b585848826f6bbaeee91fe62a4391ee5eR22... However, this may not be required if we end up installing hipify_torch for CUDA builds as well, due to the other failure wrt test_testing
2. conditionalize the use of this context manager to use it just for ROCm builds

**Comment by jithunnair-amd** [comment]
@huydhn @clee2000 
This PR is to move hipify logic to a submodule so it will involve a lot of LOC.

Do we need to do anything special to waive this requirement for this PR? Or do we merge it (when other checks are passing) with a `-f "LOC check expected to fail"` or something?

**Comment by huydhn** [comment]
> @huydhn @clee2000 This PR is to move hipify logic to a submodule so it will involve a lot of LOC.
> 
> Do we need to do anything special to waive this requirement for this PR? Or do we merge it (when other checks are passing) with a `-f "LOC check expected to fail"` or something?

You can add the label `skip-pr-sanity-checks` to the PR to bypass it. You would need to push a new commit to make that label into effect though, which should be ok when the PR is a WIP.  On the other hand, you can always use force merge when the PR is ready to merge and only that sanity check fails.

**Comment by huydhn** [comment]
Just FYI, BC linter can also be suppressed with `suppress-bc-linter` if necessary.  Same principal applies, pushing a new commit would put it into effect

**Comment by albanD** [comment]
@huydhn

>  You would need to push a new commit to make that label into effect though

you don't actually! The job will be updated once the label is added.

> On the other hand, you can always use force merge when the PR is ready to merge and only that sanity check fails.

Please let's not do that!

**Comment by lcskrishna** [comment]
@jithunnair-amd @huydhn The bc_linter seems to be failing due to removal of the torch.utils.hipify.*.py, and this PR is to remove this file and use the code present in hipify_torch. 
Can we add ```suppress-bc-linter``` label to this PR so that the CI gets fixed and we can move this PR from draft to ready-for-review?

**Comment by albanD** [comment]
@lcskrishna Actually after a short github search: https://github.com/search?q=%2Ftorch%5C.utils%5C.hipify%2F&type=code it seems like explicitly breaking this API is not ok.
Would it be possible to just point the current public API to use the new submodule so that we can preserve the public API here?

**Comment by lcskrishna** [comment]
@albanD The public API is still intact. The API points to the files present in hipify_torch (submodule) in the __init__.py file as in below, so that any other libraries using torch.utils.hipify won't have any issues.  
https://github.com/pytorch/pytorch/pull/101391/files#diff-b3ec44a93ae9dbc8f8801a9a7a41d37a781d5749d83cc6074b146dd46f2c113b

File: torch/utils/cpp_extension.py [hunk 3]
```diff
@@ -1487,8 +1489,8 @@ def _jit_compile(name,
         baton = FileBaton(os.path.join(build_directory, 'lock'))
         if baton.try_acquire():
             try:
-                with GeneratedFileCleaner(keep_intermediates=keep_intermediates) as clean_ctx:
-                    if IS_HIP_EXTENSION and (with_cuda or with_cudnn):
+                if IS_HIP_EXTENSION and (with_cuda or with_cudnn):
+                    with hipify_python.GeneratedFileCleaner(keep_intermediates=keep_intermediates) as clean_ctx:
```
Comment (albanD) [inline torch/utils/cpp_extension.py:1493]: The line above should throw if hipify_python is None.

File: torch/utils/hipify/__init__.py [hunk 4]
```diff
@@ -1 +1,6 @@
-from .version import __version__
+import hipify_torch  # type: ignore[import]
```
Comment (albanD) [inline torch/utils/hipify/__init__.py:1]: Throw a nice error here if not available?

**Review by albanD — COMMENTED** [review]
Did a more detailed review. This will still be BC-breaking for these users and they will have to add a new `pip install` for their code to still work?

**Comment by jithunnair-amd** [comment]
Unfortunately, this effort is being shelved (again).

The objective of this effort was to avoid duplication across PyTorch's hipify and hipify_torch repo. But after assessing what it would take to get the various use-cases of:
a) build from source
b) wheel install and
c) CI runs
working with this change, it doesn't seem to be worth the effort. More specifically, case a) could be handled by updating requirements.txt to pip install hipify_torch from source, while case c) could be handled by updating the .ci scripts to pip install hipify_torch from source, but case b) would require generating and publishing wheels for hipify_torch in the pytorch.org index, while still having a way to update the wheels whenever a commit bump is needed for hipify_torch. 

We will continue to conduct frequent syncs between hipify_torch and pytorch's hipify for the time being.
